pkg/apis/pipeline/v1/task_types.go

```diff
@@ -52,6 +53,61 @@ func (*Task) GetGroupVersionKind() schema.GroupVersionKind {
 	return SchemeGroupVersion.WithKind(pipeline.TaskControllerName)
 }
 
+func findWorkspaceSubstitutionLocationsInSidecars(sidecars []Sidecar) sets.String {
+	locationsToCheck := sets.NewString()
+	for _, sidecar := range sidecars {
+		locationsToCheck.Insert(sidecar.Script)
+
+		for i := range sidecar.Args {
+			locationsToCheck.Insert(sidecar.Args[i])
+		}
+
+		for i := range sidecar.Command {
+			locationsToCheck.Insert(sidecar.Command[i])
+		}
+	}
+	return locationsToCheck
+}
+
+func findWorkspaceSubstitutionLocationsInSteps(steps []Step) sets.String {
+	locationsToCheck := sets.NewString()
+	for _, step := range steps {
+		locationsToCheck.Insert(step.Script)
+
+		for i := range step.Args {
+			locationsToCheck.Insert(step.Args[i])
+		}
+
+		for i := range step.Command {
+			locationsToCheck.Insert(step.Command[i])
+		}
+	}
+	return locationsToCheck
+}
+
+func findWorkspaceSubstitutionLocationsInStepTemplate(stepTemplate *StepTemplate) sets.String {
+	locationsToCheck := sets.NewString()
+
+	if stepTemplate != nil {
+		for i := range stepTemplate.Args {
+			locationsToCheck.Insert(stepTemplate.Args[i])
+		}
+		for i := range stepTemplate.Command {
+			locationsToCheck.Insert(stepTemplate.Command[i])
+		}
+	}
+	return locationsToCheck
+}
+
+// FindWorkspaceSubstitutionLocationsInTask returns a set of all the locations in the TaskSpec  where we can apply substitutions.
+func (ts *TaskSpec) FindWorkspaceSubstitutionLocationsInTask() sets.String {
```

Comment:
Can this function be moved out of the apis package?

Reply:
I think it can. I can move it closer to where it is called.

Reply:
Would it make more sense to move it to `pkg/workspaces`?

Reply:
that could work! and instead of "workspace substitution locations", maybe it could return a list of workspace names used in the task spec

Reply:
That works too. But, to get the workspace names that are used by the task spec, I will have to parse the strings (args, command etc.) that could contain workspace substitutions so I think I will still have to extract variables names using the function defined in substitution.

The reason for doing this is that because of propagated workspaces, the workspaces might not be declared in the pipeline tasks and instead, directly be used in the steps of those tasks. We only want to propagate workspaces to this task that it needs.

Reply:
I implemented your suggestion.

Reply:
Thanks! I understand why you're exporting the function ExtractVariablesFromString in the substitution package, but I still don't understand why it's necessary to have the function `FindWorkspaceSubstitutionLocationsInTask` in the workspace pkg instead of a function like `FindWorkspaceNamesUsedInTaskSpec`. The actual string substitutions can happen in the TaskRun reconciler instead

Reply:
I think I implemented what you are suggesting. I wrote an exported function called `FindWorkspacesUsedByTask` in `pkg/workspaces` which actually returns the names of the workspaces that are used in the Task Spec. It internally uses `ExtractVariablesFromString` to parse the strings that may contain things like (`$(workspaces.foo.path)`) and extract the workspace names. Note, `ExtractVariablesFromString` is only parsing the strings, not doing the actual substitutions. The actual substitution is handled by the TaskRun reconciler.